### ISSUE TITLE
[SQLancer] internal transform error: fixpoint looped too many times

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#18409]
### What version of Materialize are you using?

v0.48.0-dev (f7027928e)

### How did you install Materialize?

Built from source

### What is the issue?

```
CREATE TABLE t0(c0 DECIMAL );
CREATE TABLE IF NOT EXISTS t1(c0 FLOAT   );
CREATE TABLE IF NOT EXISTS t2(c0 boolean , c1 DECIMAL );
CREATE TABLE IF NOT EXISTS t3(c0 INT );
SELECT ALL MIN(t3.c0) FROM t2, t3 RIGHT OUTER JOIN t1 ON (t2.c0) IN ((t2.c0) NOT IN (t2.c0)) RIGHT OUTER JOIN t0 ON ((t2.c0)<=(t2.c0)) WHERE (((t2.c0)OR(t2.c0))) IN (((('ℋcUi<')LIKE('1668187192'))) BETWEEN ((t1.c0) IS NOT NULL) AND (NOT (t2.c0))) GROUP BY t1.c0;
```
Fails with:
```
psql:fixpoint.sql:5: ERROR:  internal transform error: fixpoint looped too many times Fixpoint {
    name: "fixpoint",
    transforms: [
        PredicatePushdown {
            recursion_guard: RecursionGuard {
                depth: RefCell {
                    value: 0,
                },
                limit: 2048,
            },
        },
        NonNullable,
        ColumnKnowledge {
            recursion_guard: RecursionGuard {
                depth: RefCell {
                    value: 0,
                },
                limit: 2048,
            },
        },
        Demand {
            recursion_guard: RecursionGuard {
                depth: RefCell {
                    value: 0,
                },
                limit: 2048,
            },
        },
        FuseAndCollapse {
            transforms: [
                ProjectionExtraction,
                ProjectionLifting {
                    recursion_guard: RecursionGuard {
                        depth: RefCell {
                            value: 0,
                        },
                        limit: 2048,
                    },
                },
                Fusion,
                FlatMapToMap,
                Join,
                NormalizeLets {
                    inline_mfp: false,
                },
                Reduce,
                UnionNegateFusion,
                UnionBranchCancellation,
                NormalizeLets {
                    inline_mfp: false,
                },
                RedundantJoin {
                    recursion_guard: RecursionGuard {
                        depth: RefCell {
                            value: 0,
                        },
                        limit: 2048,
                    },
                },
                FoldConstants {
                    limit: Some(
                        10000,
                    ),
                },
            ],
        },
    ],
    limit: 100,
}; transformed relation: Return
  Project (#1)
    Reduce group_by=[#4] aggregates=[min(#3)]
      Join on=(eq(#0, #2, ((false <= NOT(#0)) AND (false >= (#4) IS NOT NULL))))
        Filter (#0) IS NOT NULL
          Get u1448
        Union
          Project (#0..=#2, #4)
            Filter (#0 = ((false <= NOT(#0)) AND (false >= (#2) IS NOT NULL)))
              Get l4
          Project (#0, #4, #5, #1)
            Join on=(eq(#0, #2, (false <= NOT(#0))) AND #1 = #3)
              Union
                Negate
                  Distinct group_by=[#0, #4]
                    Filter (#0 = (false <= NOT(#0)))
                      Get l4
                Distinct group_by=[#0, #1]
                  Filter (#0 = (false <= NOT(#0)))
                    Get l3
              Filter (#0 = (false <= NOT(#0)))
                Get l3
              Constant
                - (null, null)
With
  cte l4 =
    Join on=(#0 = #3)
      Union
        Project (#1, #0, #2)
          Get l2
        Project (#0, #4, #1)
          Join on=(#0 = #2 AND #1 = #3)
            Union
              Negate
                Distinct group_by=[#1, #2]
                  Get l2
              Distinct group_by=[#0, #1]
                Get l1
            Get l1
            Constant
              - (null)
      Filter (#0 <= #0)
        Get l3
  cte l3 =
    CrossJoin
      Filter (#0) IS NOT NULL
        Get l0
      Get u1446
  cte l2 =
    CrossJoin
      Get u1449
      Filter (#0 = (#0 != #0))
        Get l1
  cte l1 =
    CrossJoin
      Filter (#0 <= #0)
        Get l0
      Get u1447
  cte l0 =
    Distinct group_by=[#0]
      Get u1448
```

**Comment by aalexandrov** [comment]
I think @ggevay is best-positioned to debug this at the moment.

**Comment by ggevay** [comment]
I'll take a quick look again tomorrow morning with a fresh mind, but I'll probably postpone it.

TLDR: It's very complicated.

More details:

Here is a trace (where I disabled the panic): https://gist.github.com/ggevay/50c3c0047fc381b353c66bdb111c647f

It's the second fixpoint loop in `logical_optimizer`, which include `PredicatePushdown`. In various situations, `PredicatePushdown` doesn't remove a join predicate from its original place when pushing it down, but just creates a copy. E.g., [here](https://github.com/MaterializeInc/materialize/blob/eda55b5c68a3210faa8c624bcf646c1cd1180eac/src/transform/src/predicate_pushdown.rs#L719). This is a bit dangerous, because then this predicate will get pushed down again and again by the fixpoint loop. This usually doesn't cause a problem, because how this normally works is that in the first iteration `PredicatePushdown` pushes down some predicate and then some simplifications happen, then in the next iteration `PredicatePushdown` pushes down the same predicate again, but then usually some simplification removes this duplicate predicate so we get the same plan as before, and the fixpoint loop thus concludes.

But somehow the predicate being pushed down again and again doesn't work out here so well: We have a join equivalence in the plan that oscillates between two states, let's call them `simple` and `complicated`. Let's call the predicate that is being pushed down `p`. Let's call the other transforms that are in the same fixpoint loop `simplify()` The oscillation happens as follows: `simplify(simple+p)` makes `complicated`, and `simplify(complicated+p)` makes `simple`. I'll continue tomorrow morning.

Btw. I tried putting a `Fixpoint` around the `FuseAndCollapse` that is at the end of this `Fixpoint`, thinking that maybe we just need a little bit more simplification before the `PredicatePushdown` complicates stuff again, but it didn't fix the problem.

**Comment by aalexandrov** [comment]
Probably something to discuss with @mgree.

**Comment by ggevay** [comment]
Adding one more `ColumnKnowledge` after the `FuseAndCollapse` (line 452) fixes the problem!

**Comment by ggevay** [comment]
No, wait, sorry, it's not fixing it, I ran the wrong thing.
